File: application/src/main/java/com/smtm/application/security/v1/TokenController.java
package com.smtm.application.security.v1;

import java.util.Optional;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import com.smtm.security.api.Authentication;

@RestController
@RequestMapping(path = "/security/token")
public class TokenController {

    private final Authentication authentication;

    public TokenController(Authentication authentication) {
        this.authentication = authentication;
    }

    @PostMapping(
        consumes = MediaTypes.CREDENTIALS_VALUE,
        produces = MediaTypes.TOKEN_VALUE
    )
    public ResponseEntity<?> createToken(@RequestBody CredentialsDto credentialsDto) {
        return Optional.ofNullable(authentication.authenticate(credentialsDto.getEmail(), credentialsDto.getPassword()))
            .map(token -> ResponseEntity.ok(TokenDto.of(token)))
            .orElseGet(() -> ResponseEntity.status(401).build());
    }
}

File: application/src/main/kotlin/com/smtm/application/security/v1/TokenController.kt
package com.smtm.application.security.v1

import com.smtm.security.api.Authentication
import com.smtm.security.authentication.Tokens
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping(path = ["/security/tokens"])
class TokenController(private val authentication: Authentication) {

    @PostMapping(
        consumes = [CredentialsDto.MediaTypeValue],
        produces = [TokensDto.MediaTypeValue]
    )
    fun createToken(@RequestBody credentials: CredentialsDto): ResponseEntity<*> = credentials
        .authenticate()
        ?.toResponseEntity()
        ?: unauthorized

    private fun CredentialsDto.authenticate() = authentication.authenticate(email, password)

}

private fun Tokens.toResponseEntity() = tokensDtoOf(accessToken.value, refreshToken.value)
    .let { ResponseEntity.ok(it) }

private val unauthorized = ResponseEntity.status(401).build<Nothing>()
